feat: Update provider requirements to 4.0, TF version to +0.15 and devtools to v1.3 (#49)

apeabody opened the pull request.

comment-bot-dev commented:
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_python
 Failed flake8 Python lint check. 
 ```bash 
Running flake8
./06-Cloud-Function/function_source/main.py:39:58: E999 SyntaxError: invalid syntax
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
Running terraform validate
terraform_validate ./01-Getting-Started 
Success!
 The configuration is valid.
terraform_validate ./02-IAM 
There are some problems with the configuration, described below.
The Terraform configuration must be valid before initialization so that
Terraform can determine which modules and providers need to be installed.
Error: 
Missing required argument
  on main.tf line 48, in module "project_iam_bindings":
  48: module "project_iam_bindings" 
{
The argument "source" is required, but no definition was found.
terraform_validate ./03-Networking 
There are some problems with the configuration, described below.
The Terraform configuration must be valid before initialization so that
Terraform can determine which modules and providers need to be installed.
Error: 
Missing required argument
  on main.tf line 44, in module "network":
  44: module "network" 
{
The argument "source" is required, but no definition was found.
Error: 
Missing required argument
  on main.tf line 61, in module "cloud_nat":
  61: module "cloud_nat" 
{
The argument "source" is required, but no definition was found.
terraform_validate ./04-Instance-Group 
There are some problems with the configuration, described below.
The Terraform configuration must be valid before initialization so that
Terraform can determine which modules and providers need to be installed.
Error: 
Invalid resource type name
  on main.tf line 48, in resource "" "service_account_user":
  48: resource 
""
 "service_account_user" {
A name must start with a letter or underscore and may contain only letters,
digits, underscores, and dashes.
Error: 
Missing required argument
  on main.tf line 69, in module "instance_template":
  69: module "instance_template" 
{
The argument "source" is required, but no definition was found.
Error: 
Missing required argument
  on main.tf line 89, in module "managed_instance_group":
  89: module "managed_instance_group" 
{
The argument "source" is required, but no definition was found.
terraform_validate ./05-Load-Balancer 
There are some problems with the configuration, described below.
The Terraform configuration must be valid before initialization so that
Terraform can determine which modules and providers need to be installed.
Error: 
Missing required argument
  on main.tf line 82, in module "load_balancer":
  82: module "load_balancer" 
{
The argument "source" is required, but no definition was found.
terraform_validate ./06-Cloud-Function 
There are some problems with the configuration, described below.
The Terraform configuration must be valid before initialization so that
Terraform can determine which modules and providers need to be installed.
Error: 
Missing required argument
  on main.tf line 59, in module "image_processing_function":
  59: module "image_processing_function" 
{
The argument "source" is required, but no definition was found.
terraform_validate ./other/terraform-codelab/lab-app 
Warning: 
Version constraints inside provider configuration blocks are deprecated
  on provider.tf line 18, in provider "google":
  18:   version     = 
"~> 2.5"
Terraform 0.13 and earlier allowed provider version constraints inside the
provider configuration block, but that is now deprecated and will be removed
in a future version of Terraform. To silence this warning, move the provider
version constraint into the required_providers block.
Error: 
Invalid function argument
  on provider.tf line 20, in provider "google":
  20:   credentials = file("
credentials.json
")
Invalid value for "path" parameter: no file exists at credentials.json; this
function works only with files that are distributed as part of the
configuration source code, so if this file will be created by a resource in
this configuration you must instead obtain this result from an attribute of
that resource.
terraform_validate ./other/terraform-codelab/lab-networking 
Error: 
Invalid function argument
  on provider.tf line 19, in provider "google":
  19:   credentials = file("
credentials.json
")
Invalid value for "path" parameter: no file exists at credentials.json; this
function works only with files that are distributed as part of the
configuration source code, so if this file will be created by a resource in
this configuration you must instead obtain this result from an attribute of
that resource.
terraform_validate ./test/fixtures/01-Getting-Started 
Error: 
Unsupported argument
  on example.tf line 19, in module "example":
  19:   
project_id
 = var.project_id
An argument named "project_id" is not expected here.
Error: 
Unsupported argument
  on example.tf line 20, in module "example":
  20:   
region
     = "us-central1"
An argument named "region" is not expected here.
terraform_validate ./test/fixtures/02-IAM 
Error: 
Failed to download module
Could not download module "project_iam_bindings" (../../../02-IAM/main.tf:48)
source code from "": invalid source string: 
Error: 
Failed to download module
Could not download module "project_iam_bindings" (../../../02-IAM/main.tf:48)
source code from "": invalid source string: 
terraform_validate ./test/fixtures/03-Networking 
Error: 
Failed to download module
Could not download module "cloud_nat" (../../../03-Networking/main.tf:61)
source code from "": invalid source string: 
Error: 
Failed to download module
Could not download module "network" (../../../03-Networking/main.tf:44) source
code from "": invalid source string: 
Error: 
Failed to download module
Could not download module "cloud_nat" (../../../03-Networking/main.tf:61)
source code from "": invalid source string: 
Error: 
Failed to download module
Could not download module "cloud_nat" (../../../03-Networking/main.tf:61)
source code from "": invalid source string: 
Error: 
Failed to download module
Could not download module "network" (../../../03-Networking/main.tf:44) source
code from "": invalid source string: 
Error: 
Failed to download module
Could not download module "cloud_nat" (../../../03-Networking/main.tf:61)
source code from "": invalid source string: 
Error: 
Failed to download module
Could not download module "network" (../../../03-Networking/main.tf:44) source
code from "": invalid source string: 
Error: 
Failed to download module
Could not download module "cloud_nat" (../../../03-Networking/main.tf:61)
source code from "": invalid source string: 
Error: 
Failed to download module
Could not download module "network" (../../../03-Networking/main.tf:44) source
code from "": invalid source string: 
terraform_validate ./test/fixtures/04-Instance-Group 
Error: 
Failed to download module
Could not download module "instance_template"
(../../../04-Instance-Group/main.tf:69) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "managed_instance_group"
(../../../04-Instance-Group/main.tf:89) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "instance_template"
(../../../04-Instance-Group/main.tf:69) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "instance_template"
(../../../04-Instance-Group/main.tf:69) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "managed_instance_group"
(../../../04-Instance-Group/main.tf:89) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "instance_template"
(../../../04-Instance-Group/main.tf:69) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "managed_instance_group"
(../../../04-Instance-Group/main.tf:89) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "instance_template"
(../../../04-Instance-Group/main.tf:69) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "managed_instance_group"
(../../../04-Instance-Group/main.tf:89) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "instance_template"
(../../../04-Instance-Group/main.tf:69) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "managed_instance_group"
(../../../04-Instance-Group/main.tf:89) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "instance_template"
(../../../04-Instance-Group/main.tf:69) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "managed_instance_group"
(../../../04-Instance-Group/main.tf:89) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "instance_template"
(../../../04-Instance-Group/main.tf:69) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "managed_instance_group"
(../../../04-Instance-Group/main.tf:89) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "instance_template"
(../../../04-Instance-Group/main.tf:69) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "managed_instance_group"
(../../../04-Instance-Group/main.tf:89) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "instance_template"
(../../../04-Instance-Group/main.tf:69) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "managed_instance_group"
(../../../04-Instance-Group/main.tf:89) source code from "": invalid source
string: 
terraform_validate ./test/fixtures/05-Load-Balancer 
Error: 
Failed to download module
Could not download module "load_balancer"
(../../../05-Load-Balancer/main.tf:82) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "load_balancer"
(../../../05-Load-Balancer/main.tf:82) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "load_balancer"
(../../../05-Load-Balancer/main.tf:82) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "load_balancer"
(../../../05-Load-Balancer/main.tf:82) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "load_balancer"
(../../../05-Load-Balancer/main.tf:82) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "load_balancer"
(../../../05-Load-Balancer/main.tf:82) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "load_balancer"
(../../../05-Load-Balancer/main.tf:82) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "load_balancer"
(../../../05-Load-Balancer/main.tf:82) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "load_balancer"
(../../../05-Load-Balancer/main.tf:82) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "load_balancer"
(../../../05-Load-Balancer/main.tf:82) source code from "": invalid source
string: 
terraform_validate ./test/fixtures/06-Cloud-Function 
Error: 
Failed to download module
Could not download module "image_processing_function"
(../../../06-Cloud-Function/main.tf:59) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "image_processing_function"
(../../../06-Cloud-Function/main.tf:59) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "image_processing_function"
(../../../06-Cloud-Function/main.tf:59) source code from "": invalid source
string: 
Error: 
Failed to download module
Could not download module "image_processing_function"
(../../../06-Cloud-Function/main.tf:59) source code from "": invalid source
string: 
terraform_validate ./test/setup 
Success!
 The configuration is valid.
 ```

bharathkkb commented on 01-Getting-Started/main.tf:
```diff
@@ -27,8 +27,13 @@
+terraform {
+  required_providers {
+    google = {
+      version = "~> 3.53"
```
Would be great if we can bump this to the latest 4.0 provider. This may also need some associated module version bumps.

apeabody replied:
Yes, after a quick spot check it appears most module versions will need to be updated to support >3.53

comment-bot-dev commented:
@apeabody
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_headers
 All files need a license header. Please make sure all your files include the appropriate header. A helper tool available [here](https://github.com/google/addlicense). 
 ```bash 
Checking file headers
./test/integration/05-Load-Balancer/inspec.yml
./test/integration/03-Networking/inspec.yml
./test/integration/04-Instance-Group/inspec.yml
./test/integration/02-IAM/inspec.yml
./test/integration/06-Cloud-Function/inspec.yml
./test/integration/01-Getting-Started/inspec.yml
./.github/workflows/stale.yml
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
Running terraform validate
terraform_validate ./01-Getting-Started
Success!
 The configuration is valid.
terraform_validate ./02-IAM
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "project_iam_bindings" (declared at main.tf line 48) has invalid
│
 
source address "": invalid source string: .
╵
terraform_validate ./03-Networking
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "cloud_nat" (declared at main.tf line 61) has invalid source address
│
 
"": invalid source string: .
╵
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "network" (declared at main.tf line 44) has invalid source address
│
 
"": invalid source string: .
╵
terraform_validate ./04-Instance-Group
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "instance_template" (declared at main.tf line 69) has invalid source
│
 
address "": invalid source string: .
╵
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "managed_instance_group" (declared at main.tf line 89) has invalid
│
 
source address "": invalid source string: .
╵
terraform_validate ./05-Load-Balancer
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "load_balancer" (declared at main.tf line 83) has invalid source
│
 
address "": invalid source string: .
╵
terraform_validate ./06-Cloud-Function
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "image_processing_function" (declared at main.tf line 59) has
│
 
invalid source address "": invalid source string: .
╵
terraform_validate ./other/terraform-codelab/lab-app
╷
│
 
Error: 
Invalid function argument
│
 
│
 
  on provider.tf line 19, in provider "google":
│
 
  19:   credentials = file("
credentials.json
")
│
 
│
 
Invalid value for "path" parameter: no file exists at "credentials.json";
│
 
this function works only with files that are distributed as part of the
│
 
configuration source code, so if this file will be created by a resource in
│
 
this configuration you must instead obtain this result from an attribute of
│
 
that resource.
╵
terraform_validate ./other/terraform-codelab/lab-networking
╷
│
 
Error: 
Invalid function argument
│
 
│
 
  on provider.tf line 19, in provider "google":
│
 
  19:   credentials = file("
credentials.json
")
│
 
│
 
Invalid value for "path" parameter: no file exists at "credentials.json";
│
 
this function works only with files that are distributed as part of the
│
 
configuration source code, so if this file will be created by a resource in
│
 
this configuration you must instead obtain this result from an attribute of
│
 
that resource.
╵
terraform_validate ./test/fixtures/01-Getting-Started
╷
│
 
Error: 
Unsupported argument
│
 
│
 
  on example.tf line 19, in module "example":
│
 
  19:   
project_id
 = var.project_id
│
 
│
 
An argument named "project_id" is not expected here.
╵
╷
│
 
Error: 
Unsupported argument
│
 
│
 
  on example.tf line 20, in module "example":
│
 
  20:   
region
     = "us-central1"
│
 
│
 
An argument named "region" is not expected here.
╵
terraform_validate ./test/fixtures/02-IAM
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "project_iam_bindings" (declared at ../../../02-IAM/main.tf line 48)
│
 
has invalid source address "": invalid source string: .
╵
terraform_validate ./test/fixtures/03-Networking
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "cloud_nat" (declared at ../../../03-Networking/main.tf line 61) has
│
 
invalid source address "": invalid source string: .
╵
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "network" (declared at ../../../03-Networking/main.tf line 44) has
│
 
invalid source address "": invalid source string: .
╵
terraform_validate ./test/fixtures/04-Instance-Group
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "instance_template" (declared at ../../../04-Instance-Group/main.tf
│
 
line 69) has invalid source address "": invalid source string: .
╵
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "managed_instance_group" (declared at
│
 
../../../04-Instance-Group/main.tf line 89) has invalid source address "":
│
 
invalid source string: .
╵
terraform_validate ./test/fixtures/05-Load-Balancer
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "load_balancer" (declared at ../../../05-Load-Balancer/main.tf line
│
 
83) has invalid source address "": invalid source string: .
╵
terraform_validate ./test/fixtures/06-Cloud-Function
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "image_processing_function" (declared at
│
 
../../../06-Cloud-Function/main.tf line 59) has invalid source address "":
│
 
invalid source string: .
╵
terraform_validate ./test/setup
Success!
 The configuration is valid.
 ```

apeabody commented:
@bharathkkb 

* Updated Terraform provider syntax format for TF v0.12.26+ #47
* Updated Google Terraform provider to v4.0
* Updated Modules as needed for compatibility with Google provider v4.0
* Updated Labs to require Terraform v0.13+ for module compatibility, Lab 05 and tests to require Terraform v0.15+
* Updated cft/developer-tools to v1.3 for newer Terraform #49
* Updated Lab 05 for lb-http modules changes: added custom_response_headers and fix [sensitive output](https://github.com/terraform-google-modules/terraform-google-lb-http/commit/386bc8959c85e8cd06d2c3e49ffc8671763ead5e) breaking the integration test

The Linter ERROR is for a "missing" credentials file in [/other/terraform-codelab](https://github.com/terraform-google-modules/cloud-foundation-training/tree/master/other/terraform-codelab)

bharathkkb commented on 05-Load-Balancer/versions.tf:
```diff
@@ -15,5 +15,5 @@
 
 terraform {
-  required_version = ">= 0.12"
+  required_version = ">= 0.15"
```
Any reason why this requires 0.15? We should standardize on one min version.

apeabody replied:
Hi @bharathkkb The integration tests require >= 0.15 for this Lab 05 [output](https://github.com/terraform-google-modules/cloud-foundation-training/pull/48/files#diff-f841f1332ff01dce17165326ae0e0af9b1983e5630a83916b154677792bedcb4R26) due to a change in the lb-http module, however it turns out that Lab 05 itself doesn't make use of that particular output and >=0.13 would be sufficient.  Therefore all the labs would be fine with >= 0.13, but as the integration tests are running >= 0.15, I would suggest it prudent to standardize everything to >= 0.15.  Unless there is a need to support 0.13 and 0.14?

bharathkkb replied:
@apeabody gotcha, we can just standardize to `>= 0.15`. Most users leverage cloudshell for this tutorial and that also uses >= 0.15.

apeabody replied:
👍

bharathkkb reviewed (approved):
Let's also run `fix_headers` to fix the CI lint check about headers

comment-bot-dev commented:
@apeabody
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
Running terraform validate
terraform_validate ./01-Getting-Started
Success!
 The configuration is valid.
terraform_validate ./02-IAM
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "project_iam_bindings" (declared at main.tf line 48) has invalid
│
 
source address "": invalid source string: .
╵
terraform_validate ./03-Networking
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "cloud_nat" (declared at main.tf line 61) has invalid source address
│
 
"": invalid source string: .
╵
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "network" (declared at main.tf line 44) has invalid source address
│
 
"": invalid source string: .
╵
terraform_validate ./04-Instance-Group
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "instance_template" (declared at main.tf line 69) has invalid source
│
 
address "": invalid source string: .
╵
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "managed_instance_group" (declared at main.tf line 89) has invalid
│
 
source address "": invalid source string: .
╵
terraform_validate ./05-Load-Balancer
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "load_balancer" (declared at main.tf line 83) has invalid source
│
 
address "": invalid source string: .
╵
terraform_validate ./06-Cloud-Function
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "image_processing_function" (declared at main.tf line 59) has
│
 
invalid source address "": invalid source string: .
╵
terraform_validate ./other/terraform-codelab/lab-app
╷
│
 
Error: 
Invalid function argument
│
 
│
 
  on provider.tf line 19, in provider "google":
│
 
  19:   credentials = file("
credentials.json
")
│
 
│
 
Invalid value for "path" parameter: no file exists at "credentials.json";
│
 
this function works only with files that are distributed as part of the
│
 
configuration source code, so if this file will be created by a resource in
│
 
this configuration you must instead obtain this result from an attribute of
│
 
that resource.
╵
terraform_validate ./other/terraform-codelab/lab-networking
╷
│
 
Error: 
Invalid function argument
│
 
│
 
  on provider.tf line 19, in provider "google":
│
 
  19:   credentials = file("
credentials.json
")
│
 
│
 
Invalid value for "path" parameter: no file exists at "credentials.json";
│
 
this function works only with files that are distributed as part of the
│
 
configuration source code, so if this file will be created by a resource in
│
 
this configuration you must instead obtain this result from an attribute of
│
 
that resource.
╵
terraform_validate ./test/fixtures/01-Getting-Started
╷
│
 
Error: 
Unsupported argument
│
 
│
 
  on example.tf line 19, in module "example":
│
 
  19:   
project_id
 = var.project_id
│
 
│
 
An argument named "project_id" is not expected here.
╵
╷
│
 
Error: 
Unsupported argument
│
 
│
 
  on example.tf line 20, in module "example":
│
 
  20:   
region
     = "us-central1"
│
 
│
 
An argument named "region" is not expected here.
╵
terraform_validate ./test/fixtures/02-IAM
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "project_iam_bindings" (declared at ../../../02-IAM/main.tf line 48)
│
 
has invalid source address "": invalid source string: .
╵
terraform_validate ./test/fixtures/03-Networking
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "cloud_nat" (declared at ../../../03-Networking/main.tf line 61) has
│
 
invalid source address "": invalid source string: .
╵
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "network" (declared at ../../../03-Networking/main.tf line 44) has
│
 
invalid source address "": invalid source string: .
╵
terraform_validate ./test/fixtures/04-Instance-Group
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "instance_template" (declared at ../../../04-Instance-Group/main.tf
│
 
line 69) has invalid source address "": invalid source string: .
╵
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "managed_instance_group" (declared at
│
 
../../../04-Instance-Group/main.tf line 89) has invalid source address "":
│
 
invalid source string: .
╵
terraform_validate ./test/fixtures/05-Load-Balancer
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "load_balancer" (declared at ../../../05-Load-Balancer/main.tf line
│
 
83) has invalid source address "": invalid source string: .
╵
terraform_validate ./test/fixtures/06-Cloud-Function
╷
│
 
Error: 
Invalid module source address
│
 
│
 
Module "image_processing_function" (declared at
│
 
../../../06-Cloud-Function/main.tf line 59) has invalid source address "":
│
 
invalid source string: .
╵
terraform_validate ./test/setup
Success!
 The configuration is valid.
 ```
